### PR TITLE
Add net.displaycal.DisplayCAL

### PR DESCRIPTION
**Pull request by hfiguiere**
Probably need more polish / testing.

Upstrem discussion https://discuss.pixls.us/t/rip-displaycal/21775

### CODE REVIEW

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by garrett** [comment]
I haven't had much time to test it yet. Apologies!

I just tried running it on my laptop's screen a few minutes ago.

Good news: The UI shows up.

However:
1. The names of the displays aren't displayed in the dropdown (they're shown as "Monitor 1, Output XWAYLAND4 @ 0, 0, 1600x900")
2. I have some weird command line errors flash in the UI
3. an odd error on the initial check with my color calibration device

(I assume issue 3 is most likely related to 2)

displaycal

**Comment by sysrich** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by sysrich** [comment]
I've tested it on Tumbleweed with GNOME + X and it seemed to perform well, just had some issues getting the .icc profile out (fine once I digged into ~/.var though) for use

**Comment by garrett** [comment]
@sysrich Oh cool! I hope my issues above have been addressed then. I'll have to try it soon.

Meanwhile, we might want to do something special to make sure the profiles land in the right place. I guess it'd need ~/.local/share/icc/ for user specific profiles. I don't know how it handles system profiles... (I think it'd need root for that.)

**Comment by rogerxxxx** [comment]
Any way a messily user can test install this on their system, or is the DisplayCal flatpak only available for those with write access to this repository?

**Comment by hfiguiere** [comment]
The build as expired. Sorry I meant to get back to it. I'll trigger a new build if you want to test it.

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by rogerxxxx** [comment]
Was able to download, ~600 MegaBytes, wow.  Regardless, some of us need this, else I need to dig through the previous runs of displaycal and learn how to run argyll through command line interface.

**Comment by fcoiffie** [comment]
On Gentoo, I get this error :
image

**Comment by rogerxxxx** [comment]
fcoiffie: The first error states dbus was disabled when compiling xorg on your box.  Odd, I'm not seeing a dbus USE flag for xorg (packages) here.  Regardless, check to make sure dbus is not disabled within your /etc/make.conf and/or check your system profile emerge --info |grep dbus?

**Comment by hfiguiere** [comment]
I still have issue with running it with ColorHug.

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
This new build will use Wayland if available.

**Comment by hfiguiere** [comment]
> fcoiffie: The first error states dbus was disabled when compiling xorg on your box. Odd, I'm not seeing a dbus USE flag for xorg (packages) here. Regardless, check to make sure dbus is not disabled within your /etc/make.conf and/or check your system profile emerge --info |grep dbus?

The flag might be on dbus since it is dbus autolaunch.

**Comment by garrett** [comment]
While I think denying network access is a good practice overall, especially for something that's using an EOL scripting language (Python 2.x), some colorimeters do require dynamically downloading a blob to work on first run.

I have a few (old) colorimeters here. They all happen to require a download to work for the first run.

I tried this with my "X-Rite eye-one display 2" and got the following:

```
$ net.displaycal.DisplayCAL 
XDG: [Errno 2] No translation file found for domain: xdg-user-dirs
displaycal 3.8.9.3 2019-12-14T12:16:22.165507Z
   x86_64
Python 2.7.18 (default, Mar 14 2021, 22:25:16) 
[GCC 10.2.0]
Faulthandler 3.2
wxPython 4.0.7.post2 gtk3 (phoenix) wxWidgets 3.0.5
Encoding: UTF-8
File system encoding: UTF-8
Lockfile /var/home/garrett/.var/app/net.displaycal.DisplayCAL/config/DisplayCAL/DisplayCAL.lock
Existing client using port 15411
Connecting to 15411...
Connection to 127.0.0.1:15411 failed: [Errno 111] Connection refused
Starting up...
10:47:42: Debug: Failed to connect to session manager: Could not open network socket
SDL2: libSDL2-2.0.so.0
SDL2_mixer: libSDL2_mixer-2.0.so.0
Audio module: SDL 2.0
Enumerating display devices and communication ports...
/app/bin
ArgyllCMS 2.1.2
Argyll has virtual display support
...ok.
Checking video card gamma table access for display 1...
...ok.
Checking video card gamma table access for display 2...
...ok.
Checking video card gamma table access for display 3...
...ok.
Initializing GUI...

...ok.
Ready.
Setting up scripting host at 127.0.0.1:40097
Check for application update...
Check for application update failed: Connection error (-3 Temporary failure in name resolution)
Enumerating display devices and communication ports...

ArgyllCMS 2.1.2
Argyll has virtual display support
...ok.
A change in the instrument/port configuration has been detected.
Connecting to https://displaycal.net:443...
┌──────────────────────────────────────────────────────────────────────────────┐
│ -3 Temporary failure in name resolution                                      │
└──────────────────────────────────────────────────────────────────────────────┘
```

Here's what this looks like in the GUI:

Screenshot from 2021-03-18 10-56-40

I guess I should install this for the current user only? Would system-wide even work? I guess I'll try. (If so, I suppose it'd be in `~/.var/app/` somewhere instead of `~/.local/share/` or `~/.config`.)

**Comment by hfiguiere** [comment]
Ok. We can allow network.

I only have a ColorHug and couldn't even get the thing to work. Need more patches.

**Comment by hfiguiere** [comment]
We could also pre-install them in the Flatpak. Albeit I'm not sure if it is allowed.

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
This next build should have network access.

**Comment by garrett** [comment]
Oh, https://dl.flathub.org/build-repo/40531/net.displaycal.DisplayCAL.flatpakref gives me a 404. :frowning_face:

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by garrett** [comment]
> This new build will use Wayland if available.

If I go into Flatseal and turn off X, but keep Wayland, I get this:

```
XDG: [Errno 2] No translation file found for domain: xdg-user-dirs
displaycal 3.8.9.3 2019-12-14T12:16:22.165507Z
   x86_64
Python 2.7.18 (default, Mar 20 2021, 01:03:52) 
[GCC 10.2.0]
Faulthandler 3.2
wxPython 4.0.7.post2 gtk3 (phoenix) wxWidgets 3.0.5
Encoding: UTF-8
File system encoding: UTF-8
Lockfile /home/garrett/.var/app/net.displaycal.DisplayCAL/config/DisplayCAL/DisplayCAL.lock
Starting up...
Unable to access the X Display, is $DISPLAY set properly?
Ran application exit handlers
```

I'm on Wayland on Fedora 34 beta.

The displays, when I have X enabled are XWayland, so it's hard to know which is which.

**Comment by garrett** [comment]
I'm also getting this (using X, of course):

Screenshot from 2021-03-20 14-56-15

And a weird python-related error, apparently:

Screenshot from 2021-03-20 14-59-30

**Comment by hfiguiere** [comment]
> > This new build will use Wayland if available.
> 
> If I go into Flatseal and turn off X, but keep Wayland, I get this:

You need X for Argyll. Wayland is just used for Gtk3.

> The displays, when I have X enabled are XWayland, so it's hard to know which is which.
> 
> Screenshot from 2021-03-20 15-01-48
> 
> Screenshot from 2021-03-20 15-02-27

I'm tempted to say this is an XWayland issue.

Nothing much can be done on this side.

**Comment by hfiguiere** [comment]
> And a weird python-related error, apparently:
> 
> Screenshot from 2021-03-20 14-59-30

Getting this with ColorHug too.

That's a displaycal bug I need to fix. From glancing at it, it seems that it tried to find Python, fail, and the fallback is utterly broken because it doesn't handle process arguments properly.

Solving this should solve a lot of the other issues I think.

**Comment by hfiguiere** [comment]
Confirmed

tl;dr it assume python is in either `/bin` or `/usr/bin`, and since it's not return as the python executable `/usr/bin/env python`, which ultimately is wrong.

**Comment by hfiguiere** [comment]
> > > This new build will use Wayland if available.
> > 
> > 
> > If I go into Flatseal and turn off X, but keep Wayland, I get this:
> 
> You need X for Argyll. Wayland is just used for Gtk3.

DisplayCAL too, it compiles a module that uses X11 to find display info.

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
With this the calibration finished, but still can't install the profile.

**Comment by hfiguiere** [comment]
Dispwin fails with:

> Dispwin: Error - We don't have access to the VideoLUT for clearing

**Comment by rogerxxxx** [comment]
I didn't find flatpak very user intuitive for running installed applications.

I also received the DBUS error using DWM with background dbus running.

DBusException: org.freedesktop.DBus.Error.NotSupported: Using X11 for dbus-daemon autolaunch was disabled at compile time, set your DBUS_SESSION_BUS_ADDRESS instead

WORKAROUND:
$ export $(dbus-launch); flatpak run net.displaycal.DisplayCAL

**Comment by hfiguiere** [comment]
> I didn't find flatpak very user intuitive for running installed applications.

It should be in available like other applications: app menus, gnome-shell, etc.

> I also received the DBUS error using DWM with background dbus running.
> 
> DBusException: org.freedesktop.DBus.Error.NotSupported: Using X11 for dbus-daemon autolaunch was disabled at compile time, set your DBUS_SESSION_BUS_ADDRESS instead
> 
> WORKAROUND:
> $ export $(dbus-launch); flatpak run net.displaycal.DisplayCAL

Are you using Gentoo ?

**Comment by rogerxxxx** [comment]
Yes Gentoo.

**Comment by hfiguiere** [comment]
> Yes Gentoo.

So not disabling D-Bus should help then. Seem to be a pattern with Gentoo: https://github.com/flathub/flathub/pull/1999#issuecomment-796696545

**Comment by garrett** [comment]
> I didn't find flatpak very user intuitive for running installed applications.

Your app should be in all the normal places. It is on Fedora (and other distros).

Meanwhile, if you want to more easily run Flatpak apps from the command line, for whatever reasons, you can add the following to your path and the binaries would show up, even in tab-completion:

- /var/lib/flatpak/exports/bin/
- ~/.local/share/flatpak/exports/bin

(First is for system-installed flatpaks; the second is for user-installed ones.)

Apps will use the reverse naming, like `net.displaycal.DisplayCAL`, `org.gnome.Characters`, `org.mozilla.firefox`, `org.vim.Vim`, `io.neovim.nvim`, etc. but will otherwise act as a "normal" command.

**Comment by rogerxxxx** [comment]
Exactly what I was trying to explain.  Most users expect applications installed, somewhere executable via their $PATH or /usr/bin, etc...  Reverse naming is similar to Android's environment.

**Comment by hfiguiere** [comment]
> Exactly what I was trying to explain. Most users expect applications installed, somewhere executable via their $PATH or /usr/bin, etc... Reverse naming is similar to Android's environment.

Most users expect to find them in the menu (or whatever) provided by their desktop environment. Anyway, you are arguing about a Flatpak design rather than this application. This is off-topic.

**Comment by hfiguiere** [comment]
Tested with ColorHug.

File: net.displaycal.DisplayCAL.json [hunk 1]
```diff
@@ -0,0 +1,199 @@
+{
+    "id": "net.displaycal.DisplayCAL",
+    "runtime": "org.freedesktop.Platform",
+    "sdk": "org.freedesktop.Sdk//20.08",
```
Comment (barthalion) [inline net.displaycal.DisplayCAL.json:4]:
```suggestion
    "sdk": "org.freedesktop.Sdk",
```

File: net.displaycal.DisplayCAL.json [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    "runtime-version": "20.08",
+    "rename-desktop-file": "displaycal.desktop",
+    "rename-icon": "displaycal",
+    "copy-icon": true,
+    "command": "displaycal",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=x11",
```
Comment (barthalion) [inline net.displaycal.DisplayCAL.json:12]:
```suggestion
        "--socket=fallback-x11",
```

**Comment by barthalion** [comment]
bot, build net.displaycal.DisplayCAL

File: net.displaycal.DisplayCAL.json (hunk 2 above)
Comment (hadess) [reply net.displaycal.DisplayCAL.json:12]: Wasn't this already addressed in https://github.com/flathub/flathub/pull/1999#issuecomment-803381360 ?

Comment (hfiguiere) [reply net.displaycal.DisplayCAL.json:12]: we need X11. DisplayCal run external program (Argyllcms as built into the flatpak) that use X11.
Wayland here is used for the UI in Gtk3.

Comment (barthalion) [reply net.displaycal.DisplayCAL.json:12]: > Wasn't this already addressed in [#1999 (comment)](https://github.com/flathub/flathub/pull/1999#issuecomment-803381360) ?

This is open since December, I rarely look at previous reviews when submission is considered ready by authors. Also GitHub is hardly Gerrit…

@hfiguiere Noted.

**Comment by hfiguiere** [comment]
bot, build net.displaycal.DisplayCAL

File: net.displaycal.DisplayCAL.json (hunk 1 above)
Comment (hfiguiere) [reply net.displaycal.DisplayCAL.json]: done

File: net.displaycal.DisplayCAL.json (hunk 2 above)
Comment (hfiguiere) [reply net.displaycal.DisplayCAL.json:13]: added comment.
